pkg/dataobj/metastore/metastore.go

```diff
@@ -0,0 +1,175 @@
+package metastore
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/dskit/backoff"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/thanos-io/objstore"
+
+	"github.com/grafana/loki/v3/pkg/dataobj"
+	"github.com/grafana/loki/v3/pkg/logproto"
+)
+
+const (
+	metastoreWindowSize = 12 * time.Hour
+)
+
+var (
+	// Define our own builder config because metastore objects are significantly smaller.
+	metastoreBuilderCfg = dataobj.BuilderConfig{
+		SHAPrefixSize:     2,
+		TargetObjectSize:  32 * 1024 * 1024,
+		TargetPageSize:    4 * 1024 * 1024,
+		BufferSize:        32 * 1024 * 1024, // 8x page size
+		TargetSectionSize: 4 * 1024 * 1024,  // object size / 8
+	}
+)
+
+type Manager struct {
+	metastoreBuilder *dataobj.Builder
+	tenantID         string
+	metrics          *metastoreMetrics
+	bucket           objstore.Bucket
+	logger           log.Logger
+	backoff          *backoff.Backoff
+
+	builderOnce sync.Once
+}
+
+func NewMetastoreManager(bucket objstore.Bucket, tenantID string, logger log.Logger, reg prometheus.Registerer) (*Manager, error) {
+	metrics := newMetastoreMetrics()
+	if err := metrics.register(reg); err != nil {
+		return nil, err
+	}
+
+	return &Manager{
+		bucket:   bucket,
+		metrics:  metrics,
+		logger:   logger,
+		tenantID: tenantID,
+		backoff: backoff.New(context.TODO(), backoff.Config{
+			MinBackoff: 50 * time.Millisecond,
+			MaxBackoff: 10 * time.Second,
+		}),
+		builderOnce: sync.Once{},
+	}, nil
+}
+
+func (m *Manager) initBuilder() error {
+	var initErr error
+	m.builderOnce.Do(func() {
+		metastoreBuilder, err := dataobj.NewBuilder(metastoreBuilderCfg, m.bucket, m.tenantID)
+		if err != nil {
+			initErr = err
+			return
+		}
+		m.metastoreBuilder = metastoreBuilder
+	})
+	return initErr
+}
+
+func (m *Manager) UpdateMetastore(ctx context.Context, flushResult dataobj.FlushResult) error {
+	var err error
+	start := time.Now()
+	defer m.metrics.observeMetastoreProcessing(start)
+
+	// Initialize builder if this is the first call for this partition
+	if err := m.initBuilder(); err != nil {
+		return err
+	}
+
+	minTimestamp, maxTimestamp := flushResult.MinTimestamp, flushResult.MaxTimestamp
+
+	// Work our way through the metastore objects window by window, updating & creating them as needed.
+	// Each one handles its own retries in order to keep making progress in the event of a failure.
+	minMetastoreWindow := minTimestamp.Truncate(metastoreWindowSize)
+	maxMetastoreWindow := maxTimestamp.Truncate(metastoreWindowSize)
+	for metastoreWindow := minMetastoreWindow; metastoreWindow.Compare(maxMetastoreWindow) <= 0; metastoreWindow = metastoreWindow.Add(metastoreWindowSize) {
```

Comment:
`metastoreWindow.Compare(maxMetastoreWindow) <= 0` can also be written as `!metastoreWindow.After(maxMetastoreWindow)`